app/scripts/lib/transaction/decode/util.ts

```diff
@@ -0,0 +1,116 @@
+import { Hex, createProjectLogger } from '@metamask/utils';
+import EthQuery from '@metamask/eth-query';
+import {
+  DecodedTransactionDataMethod,
+  DecodedTransactionDataResponse,
+  DecodedTransactionDataSource,
+} from '../../../../../shared/types/transaction-decode';
+import { decodeUniswapRouterTransactionData } from './uniswap';
+import { decodeTransactionDataWithSourcify } from './sourcify';
+import { getContractProxyAddress } from './proxy';
+import { decodeTransactionDataWithFourByte } from './four-byte';
+
+const log = createProjectLogger('transaction-decode');
+
+export async function decodeTransactionData({
+  transactionData,
+  contractAddress,
+  chainId,
+  ethQuery,
+}: {
+  transactionData: Hex;
+  contractAddress: Hex;
+  chainId: Hex;
+  ethQuery: EthQuery;
+}): Promise<DecodedTransactionDataResponse | undefined> {
+  log('Decoding transaction data', {
+    transactionData,
+    contractAddress,
+    chainId,
+  });
+
+  const uniswapData = decodeUniswapRouterTransactionData({
+    transactionData,
+    contractAddress,
+    chainId,
+  });
+
+  if (uniswapData) {
+    log('Decoded with Uniswap commands', uniswapData);
+
+    return {
+      data: normalizeDecodedMethods(uniswapData),
+      source: DecodedTransactionDataSource.Uniswap,
+    };
+  }
+
+  const proxyAddress = await getContractProxyAddress(contractAddress, ethQuery);
+
+  if (proxyAddress) {
+    log('Retrieved proxy implementation address', proxyAddress);
+  }
+
+  const address = proxyAddress ?? contractAddress;
+
+  const sourcifyData = decodeTransactionDataWithSourcify(
+    transactionData,
+    address,
+    chainId,
+  );
+
+  const fourByteData = decodeTransactionDataWithFourByte(transactionData);
+
+  const [sourcifyResult, fourByteResult] = await Promise.allSettled([
```

Comment:
I didn't know about `Promise.allSettled()`, which is neat!

[nit] In this case we don't need to make the four byte request if sourcify successfully decodes the payload, so we can await on line 76 separately, after the return statement.

Reply:
I considered this originally, but in the pursuit of better performance, I thought it best to load both at once so the fallback was instant rather than prolonging the load.